### PR TITLE
Store `SupportedCipherSuite` in `ClientSessionValue`.

### PR DESCRIPTION
**Pull request by briansmith**
Do the mapping to `SupportedCipherSuite` at the time `ClientSessionValue` is
constructed so that we don't have to look it up all the time. This eliminates
some unwrapping.

Simplify the implementation of `find_cipher_suite` to match what is done during
`ClientSessionValue` decoding.

### CODE REVIEW

File: rustls/src/msgs/persist.rs [hunk 1]
```diff
@@ -80,9 +81,11 @@ impl Codec for ClientSessionValue {
         self.server_cert_chain.encode(bytes);
     }
 
-    fn read(r: &mut Reader) -> Option<ClientSessionValue> {
+    pub fn read(r: &mut Reader, enabled_cipher_suites: &[&'static SupportedCipherSuite]) -> Option<ClientSessionValue> {
```
Comment (briansmith) [inline rustls/src/msgs/persist.rs:84]: The prototype of `Code::read` is too restrictive. We should eventually split `Codec` into two traits: one for reading and one for encoding.

File: rustls/src/msgs/persist.rs [hunk 2]
```diff
@@ -107,22 +110,28 @@ impl Codec for ClientSessionValue {
             server_cert_chain,
         })
     }
+
+    pub fn get_encoding(&self) -> Vec<u8> {
+        let mut result = Vec::new();
+        self.encode(&mut result);
+        result
+    }
```
Comment (djc) [inline rustls/src/msgs/persist.rs:118]: This appears to be unused?

File: rustls/src/client/mod.rs [hunk 3]
```diff
@@ -433,13 +433,7 @@ impl ClientSessionImpl {
...
 
     pub fn find_cipher_suite(&self, suite: CipherSuite) -> Option<&'static SupportedCipherSuite> {
-        for scs in &self.config.ciphersuites {
-            if scs.suite == suite {
-                return Some(scs);
-            }
-        }
-
-        None
+        self.config.ciphersuites.iter().copied().find(|&scs| scs.suite == suite)
```
Comment (djc) [inline rustls/src/client/mod.rs:436]: This seems unrelated to the other changes. If the goal is to make this the same as some other code, maybe they could share an actual implementation?

File: rustls/src/msgs/persist.rs (hunk 2 above)
Comment (briansmith) [reply rustls/src/msgs/persist.rs:118]: It is used, which is why I added it. It has a default implementation in the `Codec` trait.

File: rustls/src/client/mod.rs (hunk 3 above)
Comment (briansmith) [reply rustls/src/client/mod.rs:436]: It wouldn't be right to add a dependency on `ClientSessionImpl` to the `ClientSessionValue`.

I don't think it's strictly unrelated because it helps us understand that the logic is the same. You are right that this change isn't strictly needed, so I could revert this change, if you'd prefer.

File: rustls/src/msgs/persist.rs (hunk 2 above)
Comment (briansmith) [reply rustls/src/msgs/persist.rs:118]: If you have access to the codecov report, you should be able to see it is actually executed too.

Comment (djc) [reply rustls/src/msgs/persist.rs:118]: I'm very confused. Your custom implementation appears to be identical to the default implementation -- so why is it useful to add one here?

Comment (briansmith) [reply rustls/src/msgs/persist.rs:118]: > I'm very confused. Your custom implementation appears to be identical to the default implementation -- so why is it useful to add one here?

This type no longer implements the `Codec` trait because the `Codec` trait's `read` method doesn't allow passing in additional arguments. Since it no longer implements `Codec`, it can no longer use the default implementation of `Codec::get_encoding()`.

Comment (djc) [reply rustls/src/msgs/persist.rs:118]: Ah, sorry -- the change from `impl Codec for` to `impl ` was a bit subtle.

**Review by djc — APPROVED** [review]
